### DISCUSSION TITLE
Only X MiB of Memory left

### DISCUSSION THREAD

**Discussion by bentoumia** [shadow/shadow#1843] [General]
Hello, 

first off, many thanks for the amazing piece of software!

As part of my thesis, i am currently experimenting with Shadow in the context of consensus algorithms. I am currently simulating HotStuff (https://github.com/hot-stuff/libhotstuff/) and i noticed something during the experiments with large request sizes and large amount of replicas. At some point the following warning comes up:

`00:04:38.802302 [shadow] n/a [WARN] [n/a] [manager.c:700] [_manager_checkResourceUsage] Only 310 MiB of memory available`

i took a look at memory usage as this error came up: Shadow was using about 15% of the memory, each of the replicas were using about 0.3% of memory (i had a total of 128 Replicas) and the clients were each using 2.3% (3 Clients).

My question is therefore: Is this warning a cause for concern? Would it affect the accuracy of the results or the determinism of the experiments?

Please let me know if you need me to share any of my Shadow configs.

Thank you for reading!

**Comment by stevenengler** [comment]
Shadow periodically [checks the remaining memory](https://github.com/shadow/shadow/blob/e1a0acbbc3368ed06ef6e22e28fd30343071a210/src/main/core/manager.c#L729) and logs a warning if the remaining memory drops below 500 MiB. We do this since sometimes simulations can fail with unrelated errors when the system runs out of memory, and that log message can help indicate that the failure was due to memory exhaustion. For example, the linux OOM manager may kill the managed processes when the system runs out of memory.

The warning doesn't directly mean that there's an issue, but you should keep an eye on your system memory usage. If the simulation completes successfully, then everything should be fine. Shadow itself doesn't take any special action when the memory usage is low, other than showing that warning. It's possible that some applications might independently alter their behaviour when the system memory is low, for example if the application itself monitors the memory usage and takes some action (such as shortening buffer sizes) if the available memory is low.

**Reply by bentoumia** [reply]
I understand. Many thanks for the quick reply @stevenengler!
